### PR TITLE
Consolidate DPU Host gateway initialization

### PR DESCRIPTION
**Pull request by jcaamano**
Refactor DPU Host gateway initialization to consolidate it with the other gateway modes.

### CODE REVIEW

**Comment by DmytroLinkin** [comment]
What are the benefits of this refactor? IMO, with all this additional `if`s and especially `switch` code is more spaghetti'sh.

**Comment by jcaamano** [comment]
> What are the benefits of this refactor? IMO, with all this additional `if`s and especially `switch` code is more spaghetti'sh.

A consolidated view of what we are doing for each gateway mode and less code duplication?

And because of that, there is actually much less if's with the changes than before if you count them. And the switch was already there.

**Comment by jcaamano** [comment]
Also using the gateway interface so that initialization is asynchronous just as with the other gateway modes.

**Comment by jcaamano** [comment]
I don't think `initGateway()` is significantly worse than what it was but I am open for specific suggestions. I have no problem rebasing if required after your changes.

It just makes no sense that DPU Host is initialized differently to the other gateway modes while still having so much in common.

**Comment by DmytroLinkin** [comment]
> I don't think `initGateway()` is significantly worse than what it was but I am open for specific suggestions. I have no problem rebasing if required after your changes.

Yes it doesn't. But with incremental development this "slight worsening" had accumulated over time and we have what do we have now 😄
As for the rebase, I don't see my PR will be merged soon, so yeah, no problem here)

> 
> It just makes no sense that DPU Host is initialized differently to the other gateway modes while still having so much in common.

Could be. I haven't looked into the code so deeply to check differences.

File: go-controller/pkg/node/gateway_init.go [hunk 1]
```diff
@@ -259,59 +259,86 @@ func configureSvcRouteViaInterface(iface string, gwIPs []net.IP) error {
 func (nc *DefaultNodeNetworkController) initGateway(subnets []*net.IPNet, nodeAnnotator kube.Annotator,
 	waiter *startupWaiter, managementPortConfig *managementPortConfig, kubeNodeIP net.IP) error {
 	klog.Info("Initializing Gateway Functionality")
+
+	// filter out unsupported modes
+	if config.OvnKubeNode.Mode != types.NodeModeFull && config.Gateway.Mode != config.GatewayModeShared {
```
Comment (DmytroLinkin) [inline go-controller/pkg/node/gateway_init.go:264]: Isn't it better to check this at config.go?

File: go-controller/pkg/node/gateway_init.go [hunk 2]
```diff
@@ -259,59 +259,86 @@ func configureSvcRouteViaInterface(iface string, gwIPs []net.IP) error {
...
+			return fmt.Errorf("failed to bvalidate VTEP interface MTU, err: %v", err)
+		}
+
+		if err := util.SetNodePrimaryIfAddrs(nodeAnnotator, ifAddrs); err != nil {
+			return fmt.Errorf("unable to set primary IP net label on node, err: %v", err)
+		}
+
+		if config.Gateway.EgressGWInterface != "" {
+			egressGWInterface = interfaceForEXGW(config.Gateway.EgressGWInterface)
+		}
```
Comment (DmytroLinkin) [inline go-controller/pkg/node/gateway_init.go:322]: You can move this code inside gatewayInitInternal and drop this `if` block, which will also reduce number of arguments given to the functions

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 3]
```diff
@@ -2056,3 +2056,51 @@ func addHostMACBindings(bridgeName string) error {
 	}
 	return nil
 }
+
+func newDPUHostSharedGateway(nodeName string, gwNextHops []net.IP, gwIntf string, gwIPs []*net.IPNet, watchFactory factory.NodeWatchFactory) (*gateway, error) {
+	klog.Info("Creating DPU Host Shared Gateway")
```
Comment (DmytroLinkin) [inline go-controller/pkg/node/gateway_shared_intf.go:2061]: Remove. Duplicates L333

**Review by DmytroLinkin — CHANGES_REQUESTED** [review]
Here are few suggestions. I'll return to review later, because don't like how, in general, gateway interface and interface addresses are retrieved. Basically they are used to create bridge layout. Within my PR I moved related functions into `bridgeForInterface`. With centralized gateway initialization it will be complicated now. Maybe You could look onto my PR and give some thoughts?

**Comment by trozet** [comment]
PR has gone stale. If you want you can rebase and reopen if you plan to work on it.

**Pull request closed without merging**